### PR TITLE
feat: Add GroundingDINO backend

### PR DESCRIPTION
**Pull request by shondle**
https://github.com/HumanSignal/label-studio-ml-backend/assets/106922533/470c6cfc-bb2e-4fe3-9627-30a7e3bc28d5

## GroundingDINO Backend Integration

Use text prompts for zero-shot detection of objects in images! Specify the detection of any object and get State of the Art results without any model fine tuning!

See [here](https://github.com/IDEA-Research/GroundingDINO) for more details about the pretrained GroundingDINO model.

### CODE REVIEW

**Comment by shondle** [comment]
I added Segment Anything model integration with Grounding DINO

https://github.com/HumanSignal/label-studio-ml-backend/assets/106922533/d1d2f233-d7c0-40ac-ba6f-368c3c01fd36

File: label_studio_ml/examples/grounding_dino/dino.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        
+        return predictions
+
+    def get_results(self, all_points, all_scores, all_lengths):
+        
+        results = []
+        
+        for points, scores, lengths in zip(all_points, all_scores, all_lengths):
+            # random ID
+            label_id = str(uuid4())[:4]
```
Comment (pakelley) [inline label_studio_ml/examples/grounding_dino/dino.py:129]: Why not take the whole uuid, instead of truncating it?

File: label_studio_ml/examples/grounding_dino/dino.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                    raw_img_path,
+                    label_studio_access_token=LABEL_STUDIO_ACCESS_TOKEN,
+                    label_studio_host=LABEL_STUDIO_HOST
+                )
+            except:
+                img_path = raw_img_path
+
+            src, img = load_image(img_path)
+
+            boxes, logits, phrases = predict(
```
Comment (pakelley) [inline label_studio_ml/examples/grounding_dino/dino.py:96]: nit: can we rename `phrases` to `_` to denote that it isn't used?

File: label_studio_ml/examples/grounding_dino/dino.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        lengths
+    ):
+        image = cv2.imread(img_path)
+        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+        self.predictor.set_image(image)
+
+        input_boxes = torch.from_numpy(np.array(input_boxes))
+
+        
+        transformed_boxes = transformed_boxes = self.predictor.transform.apply_boxes_torch(input_boxes, image.shape[:2])
```
Comment (pakelley) [inline label_studio_ml/examples/grounding_dino/dino.py:171]: you can get rid of one of the `transformed_boxes =` here :)

File: label_studio_ml/examples/grounding_dino/dino.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                'id': label_id,
+                'from_name': self.from_name,
+                'to_name': self.to_name,
+                'original_width': width,
+                'original_height': height,
+                'image_rotation': 0,
+                'value': {
+                    'rotation': 0,
+                    'rectanglelabels': [self.label],
+                    'width': (points[2] - points[0]) / width * 100,
```
[the remark this answers is not in the document]
Comment (shondle) [reply label_studio_ml/examples/grounding_dino/dino.py:143]: The values have to be between 0 and 100, the rest of the expression gives a decimal value I believe

https://labelstud.io/tags/rectanglelabels

File: label_studio_ml/examples/grounding_dino/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      - LOG_LEVEL=DEBUG
+
+      # Add these variables if you want to access the images stored in Label Studio
+      - LABEL_STUDIO_HOST=
+      - LABEL_STUDIO_ACCESS_TOKEN=
+
+      # use these if you want to use segment anything instead of bounding box predictions from input text prompts
+      - USE_SAM=True  # if you want to automatically generate segment anything model predictions
+      - USE_MOBILE_SAM=True # whether you want to use a more efficient, yet a bit less accurate, version of the segment anything model
+      - MOBILE_SAM_CHECKPOINT= # in case you want to point to another directory than where Docker automatically places the model
```
[the remark this answers is not in the document]
Comment (shondle) [reply label_studio_ml/examples/grounding_dino/docker-compose.yml:20]: Including SAM_CHECKPOINT now

File: label_studio_ml/examples/grounding_dino/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+RUN wget -q https://github.com/IDEA-Research/GroundingDINO/releases/download/v0.1.0-alpha/groundingdino_swint_ogc.pth
+
+WORKDIR /app
+RUN wget -q https://github.com/ChaoningZhang/MobileSAM/raw/master/weights/mobile_sam.pt
+RUN wget -q https://dl.fbaipublicfiles.com/segment_anything/sam_vit_h_4b8939.pth
+
+COPY . ./
+
+
+CMD exec gunicorn --preload --bind :$PORT --workers $WORKERS --threads $THREADS --timeout 0 _wsgi:app
```
[the remark this answers is not in the document]
Comment (shondle) [reply label_studio_ml/examples/grounding_dino/Dockerfile:33]: Not sure as this is something I copied over from some of the work @niklub did on the SAM backend Dockerfile

File: label_studio_ml/examples/grounding_dino/dino.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+import torch
+import cv2
+import matplotlib.pyplot as plt
+import numpy as np
+
+# LOADING THE MODEL
+groundingdino_model = load_model("./GroundingDINO/groundingdino/config/GroundingDINO_SwinT_OGC.py", "./GroundingDINO/weights/groundingdino_swint_ogc.pth")
+
+
+BOX_TRESHOLD = os.environ.get("BOX_THRESHOLD", 0.3)
```
[the remark this answers is not in the document]
Comment (shondle) [reply label_studio_ml/examples/grounding_dino/dino.py:23]: Adding in Dockerfile as well

File: label_studio_ml/examples/grounding_dino/dino.py (hunk 1 above)
Comment (shondle) [reply label_studio_ml/examples/grounding_dino/dino.py:129]: Not sure as this line was used from @niklub's SAM implementation which ended up fixing disappearing annotations

File: label_studio_ml/examples/grounding_dino/dino.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            'result': results
+        }]
+
+    def get_sam_results(
+        self,
+        img_path,
+        input_boxes,
+        lengths
+    ):
+        image = cv2.imread(img_path)
```
[the remark this answers is not in the document]
Comment (shondle) [reply label_studio_ml/examples/grounding_dino/dino.py:164]: load_img is a function given by DINO authors and can load straight from the path. SAM authors' loading function makes us send in a numpy array so we need to use cv2 to load it first in that case

File: label_studio_ml/examples/grounding_dino/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (pakelley) [reply label_studio_ml/examples/grounding_dino/Dockerfile:33]: gotcha, do you mind testing without it? I'll spin this up too, but since you already have everything set up you may be quicker 😅

File: label_studio_ml/examples/grounding_dino/dino.py (hunk 1 above)
Comment (niklub) [reply label_studio_ml/examples/grounding_dino/dino.py:129]: @pakelley we may better check this with @hlomzik but to the best of my knowledge, a per-list unique `id` is a requirement for interactive predictions. Truncating it only for the sake of the payload size

File: label_studio_ml/examples/grounding_dino/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (niklub) [reply label_studio_ml/examples/grounding_dino/Dockerfile:33]: yeah, I think it is a copypaste from some older versions

File: label_studio_ml/examples/grounding_dino/dino.py (hunk 1 above)
Comment (pakelley) [reply label_studio_ml/examples/grounding_dino/dino.py:129]: Gotcha yeah, my concern is mainly on how unique we can expect 4 characters to be. Do these label ids need to be unique only within their project, or across the entire database?
Honestly, either way 4 characters seems pretty low. uuid4 output is hexadecimal, so that gives us 16^4 or 65,536 possibilities. Guaranteed conflict above that, but likely conflict after... a few hundred? (rough birthday problem logic)

Comment (pakelley) [reply label_studio_ml/examples/grounding_dino/dino.py:129]: I'd recommend we use at least up to the first "-" if this only has to be unique within the project (so 8 digits, ~400M possibilities, likely collision not until over 1M)
If it has to be global, I'd recommend up to the second "-" (12 digits, not including the first "-", 300T possibilities, likely collision not until many millions of records)
All that said, maybe I'm just missing that there really aren't that many ids being generated here?

Comment (hlomzik) [reply label_studio_ml/examples/grounding_dino/dino.py:129]: ids are unique per task (annotation in our terms), so don't worry about the whole database =)

but possibilities spread over the number of tasks, so if you have 4 digits (1/65k chance), have only 10 regions in every task, but have 100k of tasks, you'll definitelly get collision somewhere.

Comment (shondle) [reply label_studio_ml/examples/grounding_dino/dino.py:140]: Ok, I'll change this to truncate at 8 digits

File: label_studio_ml/examples/grounding_dino/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (shondle) [reply label_studio_ml/examples/grounding_dino/Dockerfile]: Just removed exec

**Review by farioas — DISMISSED** [review]
Something wrong in .github folder. It shows changes that already merged. Rebase the change and push it again.

**Review by pakelley — APPROVED** [review]
